FIR checker: report UNINITIALIZED_ENUM_(ENTRY|COMPANION)

jsjeon opened the pull request:
In an `enum` class, member property initializers, enum entry, and companion object are initialized in order. That means, an access from preceding one may cause runtime exception if not alerted.

Three notable examples from code comments:
```kt
// From KT-11769
enum class Fruit(...) {
  APPLE(...);
  companion object {
    val common = ...
  }
  val score = ... <!>common<!> // UNINITIALIZED_VARIABLE
}
```

```kt
enum class EnumCompanion2(...) {
   INSTANCE(<!>foo<!>()) // UNINITIALIZED_ENUM_COMPANION
   companion object {
     fun foo = ...
   }
}
```

```kt
// From KT-6054
enum class MyEnum {
  A, B;
  val x = when(this) {
    <!>A<!> -> ... // UNINITIALIZED_ENUM_ENTRY
    <!>B<!> -> ... // ditto
  }
}
```

The checker is visiting a qualified access expression; makes sure the referenced declaration is either in an enum class or in an enum companion; and examines that container and the accessed context to determine uninitialized cases.

punzki commented on compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUninitializedEnumChecker.kt:
```diff
@@ -0,0 +1,124 @@
+            //   D1(<!>D2<!>),
+            //   D2(...)
+            // }
+            accessedContext.correspondingEnumEntry(enumEntries)?.let { caller ->
```
If you have `correspondingEnumEntry` return something like an `indexOf` (-1 if not in `enumEntries`), then in here you can just loop backwards from the index of the caller until you find the callee (or not). Or you could add some comments below; it was confusing to me.

jsjeon replied:
Hm, looping backwards sounds more confusing to me, e.g., if the other one does appear in front of the starting point. Added comments about how `indexOf` can be used here, as well as why I chose this approach.

Basically, for the given two elements of interest, iterating the list and finding any one that precedes. This way, we can iterate just once, up to that first appearance.

punzki commented on compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUninitializedEnumChecker.kt:
```diff
@@ -0,0 +1,124 @@
+        //   enum class EnumCompanion3(...) {
+        //     INSTANCE(EnumCompanion2.foo())
+        //   }
+        // If we only compare entry name, INSTANCE, we may incorrectly return EnumCompanion2.INSTANCE for EnumCompanion3.INSTANCE init.
```
To avoid this altogether, can we check if the accessed context is within the same enum class as the callee?

jsjeon replied:
Yeah, that is added to handle a case like an anonymous function you came up with, and this part is no longer necessary.

jsjeon commented on compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUninitializedEnumChecker.kt:
```diff
@@ -0,0 +1,124 @@
+
+        val accessedContext = context.containingDeclarations.lastOrNull() ?: return
+
+        // Initialization order: member property initializers, enum entries, companion object (including members in it).
```
It's about JVM class loading, how `enum` class is converted down to bytecode, etc., and I can't find a good single reference. I explain that in code comments.

punzki replied:
Very nicely written comment 🙂

jsjeon commented on compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUninitializedEnumChecker.kt:
```diff
@@ -0,0 +1,124 @@
+
+        val enumClass = if (isInsideEnum) containingClass else containingClass.outerClass(context) as? FirRegularClass ?: return
+
+        val accessedContext = context.containingDeclarations.lastOrNull() ?: return
```
Nice catch! Yeah, it should look up until the declaration is in the same enum class of interest.

jsjeon commented on compiler/testData/diagnostics/tests/controlFlowAnalysis/enumCompanionInterdependence.fir.kt:
```diff
@@ -26,21 +26,21 @@ enum class Fruit(personal: Int) {
-    INSTANCE(Companion.foo()),
-    ANOTHER(foo());
+    INSTANCE(<!UNINITIALIZED_ENUM_COMPANION!>Companion<!>.foo()),
+    ANOTHER(<!UNINITIALIZED_ENUM_COMPANION!>foo<!>());
```
Thank you for calling this out on behalf of me. Note that this is exactly same as 8 lines below, and I'm pretty much sure FE 1.0 missed this somehow.

punzki replied:
No problem. I guess we have found 3 bugs in FE1.0 in this review! 😄

tgeng commented on compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUninitializedEnumChecker.kt:
```diff
@@ -0,0 +1,166 @@
+                            context
+                        )
+                    } else {
+                        reporter.reportOn(source, FirErrors.UNINITIALIZED_ENUM_COMPANION, enumClass.symbol, context)
```
Isn't the source pointing to the entire call in this case? Should it match the declared element type (`KtSimpleNameExpression`) of `UNINITIALIZED_ENUM_COMPANION`?

jsjeon replied:
Yes, it will point to the entire call, and thus I'm using `PositioningStrategy.REFERENCE_BY_QUALIFIED` for this diagnostic in order to nail down to reference expression.

tgeng replied:
The reported element type here must match that declared in `FirErrors`. Otherwise, you would get the following class cast exception in the IDE after https://github.com/JetBrains/kotlin/pull/4250 is merged.

```
java.lang.ClassCastException: org.jetbrains.kotlin.psi.KtCallExpression cannot be cast to org.jetbrains.kotlin.psi.KtSimpleNameExpression
	at org.jetbrains.kotlin.idea.frontend.api.fir.diagnostics.UninitializedEnumCompanionImpl.getPsi(KtFirDiagnosticsImpl.kt:1520)
	at org.jetbrains.kotlin.idea.frontend.api.fir.diagnostics.UninitializedEnumCompanionImpl.getPsi(KtFirDiagnosticsImpl.kt:1520)
	at org.jetbrains.kotlin.idea.fir.highlighter.KotlinHighLevelDiagnosticHighlightingPass.addDiagnostic(KotlinHighLevelDiagnosticHighlightingPass.kt:55)
	at org.jetbrains.kotlin.idea.fir.highlighter.KotlinHighLevelDiagnosticHighlightingPass.doCollectInformation(KotlinHighLevelDiagnosticHighlightingPass.kt:48)
	at com.intellij.codeHighlighting.TextEditorHighlightingPass.collectInformation(TextEditorHighlightingPass.java:54)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$1(PassExecutorService.java:399)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1110)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$2(PassExecutorService.java:392)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:629)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:581)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.doRun(PassExecutorService.java:391)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$run$0(PassExecutorService.java:367)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:170)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:182)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.run(PassExecutorService.java:365)
	at com.intellij.concurrency.JobLauncherImpl$VoidForkJoinTask$1.exec(JobLauncherImpl.java:181)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

jsjeon replied:
You're right, thanks! Changed to `calleeReference`'s source, which is of `KtNameReferenceExpression`, which is a subtype of `KtSimpleNameExpression`.

punzki commented on compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUninitializedEnumChecker.kt:
```diff
@@ -0,0 +1,166 @@
+            //   D2(...)
+            // }
+            if (accessedContext in enumEntries && context.containingDeclarations.lastOrNull()?.isEnumEntryInitializer == true) {
+                // Technically, this is equal to `enumEntries.indexOf(accessedContext) < enumEntries.indexOf(calleeDeclaration)`.
```
I believe this should be `<=` (in the index comparison)

jsjeon replied:
Ah, self reference, thanks!

punzki commented on compiler/testData/diagnostics/tests/controlFlowAnalysis/enumInterdependence.fir.kt:
```diff
@@ -1,10 +1,10 @@
 
 enum class D(val x: Int) {
-    D1(D2.x),
+    D1(<!UNINITIALIZED_ENUM_ENTRY!>D2<!>.x),
```
Here's another case which I think is missed in FE1.0 (causes NPE) but I believe you catch (`calleeEnumEntry == accessedContext`). Could you add and try:

```kt
enum class E(val x: Int) {
  D1(D1.x);
}
```

jsjeon replied:
Indeed! Added a self-reference case.

tgeng commented:
Just want to bring attention to resolved conversation here https://github.com/JetBrains/kotlin/pull/4244#discussion_r604306388

jsjeon commented:
> Just want to bring attention to resolved conversation here [#4244 (comment)](https://github.com/JetBrains/kotlin/pull/4244#discussion_r604306388)

Thanks. Unresolved, changed the source accordingly, and resolved the comment again. :)

mglukhikh commented on compiler/fir/analysis-tests/testData/resolve/fromBuilder/enums.kt:
```diff
@@ -23,7 +23,7 @@ enum class Planet(val m: Double, internal val r: Double) {
     };
 
-    val g: Double = G * m / (r * r)
+    val g: Double = <!UNINITIALIZED_VARIABLE!>G<!> * m / (r * r)
```
Why so? It's `const val` declared in companion object.

demiurg906 replied:
Actually this is correct (FE 1.0 reports error here)

Imagine that `G` is marked with `@JvmStatic`

mglukhikh replied:
Yes, you are right.

mglukhikh commented on compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/FirHelpers.kt:
```diff
@@ -156,6 +157,17 @@ inline fun <reified T : Any> FirQualifiedAccessExpression.getDeclaration(): T? {
 
+fun FirClassLikeSymbol<*>.outerClass(context: CheckerContext): FirClassLikeSymbol<*>? {
+    if (this !is FirClassSymbol<*>) return null
+    if (classId.isLocal) return null
```
Why can we write so? I think with local class we should be able to find its outer class in checker context.

jsjeon replied:
Oh, I don't know why I added that, thanks!

mglukhikh commented on compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUninitializedEnumChecker.kt:
```diff
@@ -0,0 +1,172 @@
+    // See related discussions:
+    // https://youtrack.jetbrains.com/issue/KT-6054
+    // https://youtrack.jetbrains.com/issue/KT-11769
+    override fun check(expression: FirQualifiedAccessExpression, context: CheckerContext, reporter: DiagnosticReporter) {
```
Note that FE 1.0 performs this analysis as a part of CFA. It's a bit strange for me to see here separate logic about preceding enum entries and so. May be it's wrong architecturally.

@demiurg906 , what do you think about it?

demiurg906 replied:
Actually I like that this checker is not part of CFA, because there is no such thing as graph of static initialization in FIR CFA, and this checker exactly about it. I think that it may be moved to CFA after we add true statics to language, but for now this is only one case of such errors, so it's better to have small accurate separate checker instead of adding hacks to CFA or building graph for static sections without proper design

mglukhikh replied:
Ok

mglukhikh commented:
Sorry but a set of false positives detected in FIR bootstrap
```
    file or directory 'Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\main\kotlin', not found
09:14:16
  e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (410, 20): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
09:14:16
    file or directory 'Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\main\kotlin', not found
09:14:16
    file or directory 'Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\main\kotlin', not found
09:14:16
  e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (416, 20): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
09:14:16
    file or directory 'Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\main\kotlin', not found
09:14:16
    Using Kotlin/JVM incremental compilation
09:14:16
    i: found daemon on port 17355 (234634 ms old), trying to connect
09:14:16
  e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (417, 24): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
09:14:16
    i: connected to the daemon
09:14:16
  e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (418, 24): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
09:14:16
  e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (419, 24): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
09:14:16
  e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (426, 20): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
09:14:16
    Options for KOTLIN DAEMON: IncrementalCompilationOptions(super=CompilationOptions(compilerMode=INCREMENTAL_COMPILER, targetPlatform=JVM, reportCategories=[0], reportSeverity=2, requestedCompilationResults=[0]kotlinScriptExtensions=[kt, kts, java]), areFileChangesKnown=false, modifiedFiles=null, deletedFiles=null, workingDir=Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\build\kotlin\compileKotlin, multiModuleICSettings=MultiModuleICSettings(buildHistoryFile=Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\build\kotlin\compileKotlin\build-history.bin, useModuleDetection=false), usePreciseJavaTracking=trueoutputFiles=[Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\build\classes\kotlin\main, Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\build\kotlin\compileKotlin])
09:14:16
    Detected overflow for Z:\buildAgent\work\928586c439f50b8c\kotlin
09:14:16
    Not watching anything anymore
09:14:16
  e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (432, 20): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
09:14:16
    Watching 0 directory hierarchies to track changes
09:14:16
  e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (433, 24): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
09:14:16
    w: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\idea-frontend-fir\src\org\jetbrains\kotlin\idea\asJava\codeReferences\FirLightPsiJavaCodeReferenceElementWithReference.kt: (10, 10): org.jetbrains.kotlin.fir.declarations.impl.FirSimpleFunctionImpl@1e7422b4, org.jetbrains.kotlin.fir.declarations.impl.FirSimpleFunctionImpl@218291a9
09:14:16
    Watching 1 directory hierarchies to track changes
09:14:17
    Stored cache entry for task ':idea:idea-frontend-fir:compileKotlin' with cache key 763a3a38a0b75f928e746342adfcd621
09:14:16
  e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (434, 24): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
09:14:17
    :idea:idea-frontend-fir:compileKotlin (Thread[Execution worker for ':',5,main]) completed. Took 36.818 secs.
```

jsjeon commented:
> Sorry but a set of false positives detected in FIR bootstrap
> 
> ```
>     file or directory 'Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\main\kotlin', not found
>   e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (410, 20): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
>   e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (416, 20): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
>   e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (417, 24): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
>   e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (418, 24): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
>   e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (419, 24): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
>   e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (426, 20): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
>   e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (432, 20): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
>   e: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (433, 24): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
>     w: Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\idea-frontend-fir\src\org\jetbrains\kotlin\idea\asJava\codeReferences\FirLightPsiJavaCodeReferenceElementWithReference.kt: (10, 10): org.jetbrains.kotlin.fir.declarations.impl.FirSimpleFunctionImpl@1e7422b4, org.jetbrains.kotlin.fir.declarations.impl.FirSimpleFunctionImpl@218291a9
> Z:\buildAgent\work\928586c439f50b8c\kotlin\idea\src\org\jetbrains\kotlin\idea\refactoring\introduce\extractionEngine\ExtractableCodeDescriptor.kt: (434, 24): Companion object of enum class 'org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractionTarget' is uninitialized here
> ```

https://github.com/JetBrains/kotlin/blob/master/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/extractionEngine/ExtractableCodeDescriptor.kt#L403

Interesting way to kind of delegate/share logics to enum companion object. I'll simplify the case and will filter such case out.

jsjeon commented:
To handle an enum member property with lambda initialization (as suggested at https://github.com/JetBrains/kotlin/pull/4244#discussion_r603010409), the look-up for accessed context goes up to either enum entries or enum members. For that false positives, such lookup landed on an enum entry, but the actual, last containing declaration is _member_ function of that entry, which is legitimate, i.e., companion object will be initialized at that point. So, initialization order matters only for enum entry initializers, not enum entry members. I added a proper bail-out for that.

While running full pipeline tests, I found one more false positive: enum member property with lazy delegation where enum entry is accessed in that `lazy`. Added a simple regression test as well as proper bail-out.

mglukhikh commented:
Merged manually 👍
